File: day01/day01.hs
import System.IO (readFile)
import Data.List (lines)
import Data.List (maximumBy)
import Data.Ord (comparing)

-- | Parse a list of numbers in groups and add up the 3 largest sums of the numbers in each group.
--
-- The input is a list of strings, where each string represents a line in the input file.
-- Each line can either be empty (indicating the start of a new group of numbers),
-- or it can contain a number (which will be added to the current group of numbers).
--
-- The output is the sum of the 3 largest sums of the numbers in each group.
parseAndSum :: [String] -> Int
parseAndSum lines = sum $ take 3 $ sortBy (comparing Down) $ map sum groups
  where
    groups = foldr addNumberToGroup [[]] lines

-- | Add a number to the current group of numbers.
--
-- If the number is empty, start a new group of numbers.
-- Otherwise, add the number to the current group of numbers.
addNumberToGroup :: String -> [[Int]] -> [[Int]]
addNumberToGroup line groups
    | null line  = [] : groups
    | otherwise  = (read line : head groups) : tail groups

-- | Read the input from a file and parse it
main :: IO ()
main = do
    -- Read the contents of the file into a string
    input <- readFile "test.txt"

    -- Split the string into a list of lines
    let lines = lines input

    -- Parse the input and compute the sum of the 3 largest sums
    let totalSum = parseAndSum lines

    -- Print the result
    putStrLn $ "The total sum of the 3 largest sums is: " ++ show totalSum

File: day01/day01.py
import heapq

# Open the file
with open('input.txt', 'r') as f:
    # Initialize the list of numbers
    numbers = [[]]

    # Iterate over the lines in the file
    for line in f:
        # Strip leading and trailing whitespace from the line
        line = line.strip()

        # If the line is empty, start a new group of numbers
        if not line:
            numbers.append([])
            continue

        # Otherwise, convert the line to a number and add it to the current group of numbers
        numbers[-1].append(int(line))

# Use the map function to apply the sum function to each group of numbers
sums = map(sum, numbers)

# Convert the iterator to a list
sums = list(sums)

# Find the 3 largest elements in the list of sums
largest_sums = heapq.nlargest(3, sums)

# Add up the 3 largest sums
total_sum = sum(largest_sums)

# Print the total sum
print("The total sum of the 3 largest sums is:", total_sum)
